plugins/redfish/fu-redfish-smc-device.c

```diff
@@ -0,0 +1,217 @@
+	msgid = json_object_get_string_member(tmp_obj, "MessageId");
+	if (g_strcmp0(msgid, "SMC.1.0.OemSimpleupdateAcceptedMessage") != 0)
+		return FALSE;
+	if (!(tmp_ary = json_object_get_array_member(tmp_obj, "MessageArgs")))
+		return FALSE;
+	if (json_array_get_length(tmp_ary) != 1)
+		return FALSE;
+	return json_array_get_string_element(tmp_ary, 0);
+}
+
+static GString *
+fu_redfish_smc_device_get_parameters(FuRedfishSmcDevice *self)
+{
+	g_autoptr(GString) str = g_string_new(NULL);
+	g_autoptr(JsonBuilder) builder = json_builder_new();
+	g_autoptr(JsonGenerator) json_generator = json_generator_new();
+	g_autoptr(JsonNode) json_root = NULL;
+
+	/* create header */
+	/* https://supermicro.com/manuals/other/RedishRefGuide.pdf */
+	json_builder_begin_object(builder);
+	json_builder_set_member_name(builder, "Targets");
+	json_builder_begin_array(builder);
+	json_builder_add_string_value(builder, "/redfish/v1/Systems/1/Bios");
+	json_builder_end_array(builder);
+	json_builder_set_member_name(builder, "@Redfish.OperationApplyTime");
+	json_builder_add_string_value(builder, "OnStartUpdateRequest");
+	json_builder_set_member_name(builder, "Oem");
+	json_builder_begin_object(builder);
+	json_builder_set_member_name(builder, "Supermicro");
+	json_builder_begin_object(builder);
+	json_builder_set_member_name(builder, "BIOS");
+	json_builder_begin_object(builder);
+	json_builder_set_member_name(builder, "PreserveME");
+	json_builder_add_boolean_value(builder, TRUE);
+	json_builder_set_member_name(builder, "PreserveNVRAM");
+	json_builder_add_boolean_value(builder, TRUE);
+	json_builder_set_member_name(builder, "PreserveSMBIOS");
+	json_builder_add_boolean_value(builder, TRUE);
+	json_builder_set_member_name(builder, "BackupBIOS");
+	json_builder_add_boolean_value(builder, FALSE);
+	json_builder_end_object(builder);
+	json_builder_end_object(builder);
+	json_builder_end_object(builder);
+	json_builder_end_object(builder);
+
+	/* export as a string */
+	json_root = json_builder_get_root(builder);
+	json_generator_set_pretty(json_generator, TRUE);
+	json_generator_set_root(json_generator, json_root);
+	json_generator_to_gstring(json_generator, str);
+
+	return g_steal_pointer(&str);
+}
+
+static gboolean
+fu_redfish_smc_device_start_update(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuRedfishBackend *backend = fu_redfish_device_get_backend(FU_REDFISH_DEVICE(device));
+	JsonObject *json_obj;
+	CURL *curl;
+	const gchar *location = NULL;
+	g_autoptr(FuRedfishRequest) request = fu_redfish_backend_request_new(backend);
+
+	curl = fu_redfish_request_get_curl(request);
+	(void)curl_easy_setopt(curl, CURLOPT_POSTFIELDS, "");
+
+	if (!fu_redfish_request_perform(
+		request,
+    fu_redfish_backend_get_push_uri_path(backend),
+		FU_REDFISH_REQUEST_PERFORM_FLAG_LOAD_JSON,
+		error))
+		return FALSE;
+	json_obj = fu_redfish_request_get_json_object(request);
+
+	if (!(location = fu_redfish_smc_device_get_task(json_obj))) {
```

Comment:
i think spitting this up would be much more readable; e.g. 

    location = fu_redfish_smc_device_get_task(json_obj)
    if (location == NULL) {